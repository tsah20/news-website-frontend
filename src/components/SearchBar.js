import React from "react";

import { IconButton, InputBase, Paper } from "@material-ui/core/";
import { makeStyles } from "@material-ui/core/styles";
import SearchIcon from "@material-ui/icons/Search";

const useStyles = makeStyles((theme) => ({
  root: {
    padding: "4px 4px",
    display: "flex",
    alignItems: "center",
    width: "80%",
    marginTop: "2%",
    margin: "auto",
  },
  input: {
    marginLeft: theme.spacing(1),
    flex: 1,
    borderColor: "green",
  },
}));

/**
 *  Component Search Bar
 *  Renders the name/title of the webpage
 *  @param onSubmit Callback function for the submit
 *  @param ref for the input element
 */
export const SearchBar = React.forwardRef(({ onSubmit }, ref) => {
  const classes = useStyles();

  const handleChange = (event) => {
    event.preventDefault();
  };
  return (
    <React.Fragment>
      <Paper component="form" className={classes.root}>
        <InputBase
          onChange={handleChange}
          inputRef={ref}
          className={classes.input}
          fullWidth={true}
          placeholder="Search News with keywords"
          inputProps={{ "aria-label": "search news" }}
        />
        <IconButton
          type="submit"
          onClick={onSubmit}
          className={classes.iconButton}
          aria-label="search"
        >
          <SearchIcon />
        </IconButton>
      </Paper>
    </React.Fragment>
  );
});
