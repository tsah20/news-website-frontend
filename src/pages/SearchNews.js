import React, { useState, useRef } from "react";

import { makeStyles } from "@material-ui/core/styles";
import { useMediaQuery, CircularProgress } from "@material-ui/core/";
import Pagination from "@material-ui/lab/Pagination";

import useHttp from "../hooks/use-http";

import theme from "../theme/Theme.js";
import NoResultFound from "./NoResultFound";
import { SearchBar } from "../components/SearchBar";
import Footer from "../components/Footer";
import NewsListIterator from "../helpers/NewsListIterator";

const useStyles = makeStyles((theme) => ({
  root: {
    padding: "4px 4px",
    display: "flex",
    alignItems: "center",
    width: "80%",
    marginTop: "2%",
    margin: "auto",
  },
  input: {
    marginLeft: theme.spacing(1),
    flex: 1,
    borderColor: "green",
  },
  loader: {
    marginLeft: "60%",
  },
  pagination: {
    marginTop: "2%",
    marginRight: "10%",
    float: "right",
  },
  emptyContainerFoot: {
    position: "fixed",
  },

  postContainerFoot: {
    position: "static",
  },
}));

/* Generate the Search URL */
const getSearchUrl = (searchTerm, pageNumber) =>
  `http://localhost:5000/search?q=${searchTerm}&page=${pageNumber}`;

/**
 *  @Page Home Page shown at the root URL
 *  Home Page for the application
 *  Renders the search bar to display the news
 */
export default function SearchNews() {
  const classes = useStyles();
  const mobileContent = useMediaQuery(theme.breakpoints.down("sm"));

  const [searchNews, setSearchNews] = useState({
    totalResults: -1,
    page: "1",
    pageSize: "5",
    articles: [],
  });
  const [pages, setPages] = useState(1);
  const inputRef = useRef();
  const { isLoading, error, sendRequest: fetchNews } = useHttp();

  const handleSearch = (event, page = 1) => {
    event.preventDefault();
    const searchKey = inputRef.curent.value;
    const updateNewsData = (data) => {
      setSearchNews(data);
      /* Free API can only fetch 100 news**/
      const updatedPageCount =
        data.totalResults > 100
          ? Math.round(100 / 5)
          : Math.round(data.totalResults / 5);
      setPages(updatedPageCount);
    };
    fetchNews({ url: getSearchUrl(searchKey, page) }, updateNewsData);
  };

  const handlePageChange = (event, pageNumber) => {
    event.preventDefault();
    handleSearch(event, pageNumber);
  };

  /* Dynamically making footer float or fixed **/
  const footerCssClass =
    !isLoading && searchNews.articles.length
      ? classes.postContainerFoot
      : classes.emptyContainerFoot;

  return (
    <React.Fragment>
      <SearchBar
        data-test="component-searchbar"
        ref={inputRef}
        onSubmit={handleSearch}
      />

      {/* show circular progress bar while the content is beingloaded*/}
      {isLoading && <CircularProgress className={classes.loader} />}

      {/* show no Result when total result count is 0*/}
      {!isLoading && searchNews.totalResults == 0 && <NoResultFound />}

      {/* show content and pagination when API respinse has articles*/}
      {!isLoading && searchNews.articles.length > 0 && (
        <React.Fragment>
          <NewsListIterator newsList={searchNews.articles} />
          <Pagination
            count={pages}
            onChange={handlePageChange}
            defaultPage={1}
            className={classes.pagination}
            color="primary"
            size="large"
            showFirstButton
            showLastButton
          />
        </React.Fragment>
      )}
      <Footer customClass={footerCssClass} />
    </React.Fragment>
  );
}
