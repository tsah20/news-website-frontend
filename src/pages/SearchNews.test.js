import React from "react";
import Enzyme, { shallow, mount } from "enzyme";
import EnzymeAdapter from "@wojtekmaj/enzyme-adapter-react-17";
import { createMount, createShallow } from "@material-ui/core/test-utils";

import SearchNews from "./SearchNews";
import SearchBar from "../components/SearchBar";

Enzyme.configure({ adapter: new EnzymeAdapter() });

const setup = () => shallow(<SearchNews />);

const findByTestAttr = (wrapper, val) => wrapper.find(`[data-test='${val}']`);

describe("Search Component in intial startup", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = setup();
  });
  test("render without crashing", () => {});

  test("render non empty component without crashing", () => {
    expect(wrapper.exists()).toBe(true);
  });

  test("should match the snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
