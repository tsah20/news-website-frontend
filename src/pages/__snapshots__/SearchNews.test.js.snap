// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Component in intial startup should match the snapshot 1`] = `"<form class=\\"MuiPaper-root makeStyles-root-7 MuiPaper-elevation1 MuiPaper-rounded\\"><div class=\\"MuiInputBase-root makeStyles-input-8 MuiInputBase-fullWidth\\"><input type=\\"text\\" placeholder=\\"Search News with keywords\\" aria-label=\\"search news\\" class=\\"MuiInputBase-input\\"/></div><button class=\\"MuiButtonBase-root MuiIconButton-root\\" tabindex=\\"0\\" type=\\"submit\\" aria-label=\\"search\\"><span class=\\"MuiIconButton-label\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z\\"></path></svg></span></button></form><footer class=\\"makeStyles-footer-9 makeStyles-emptyContainerFoot-5\\"><div class=\\"MuiBox-root MuiBox-root-10\\"><div class=\\"MuiContainer-root MuiContainer-maxWidthLg\\"><div class=\\"MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-5\\"><div class=\\"MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-4\\"><div class=\\"MuiBox-root MuiBox-root-11\\">Header 1</div><div class=\\"MuiBox-root MuiBox-root-12\\"><a class=\\"MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorInherit\\" href=\\"/\\">Link 1</a></div><div class=\\"MuiBox-root MuiBox-root-13\\"><a class=\\"MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorInherit\\" href=\\"/\\">Link 2</a></div><div class=\\"MuiBox-root MuiBox-root-14\\"><a class=\\"MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorInherit\\" href=\\"/\\">Link 3</a></div></div><div class=\\"MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-4\\"><div class=\\"MuiBox-root MuiBox-root-15\\">Header 2</div><div class=\\"MuiBox-root MuiBox-root-16\\"><a class=\\"MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorInherit\\" href=\\"/\\">Link 1</a></div><div class=\\"MuiBox-root MuiBox-root-17\\"><a class=\\"MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorInherit\\" href=\\"/\\">Link 2</a></div><div class=\\"MuiBox-root MuiBox-root-18\\"><a class=\\"MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorInherit\\" href=\\"/\\">Link 3</a></div></div><div class=\\"MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-4\\"><div class=\\"MuiBox-root MuiBox-root-19\\">Header 3</div><div class=\\"MuiBox-root MuiBox-root-20\\"><a class=\\"MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorInherit\\" href=\\"/\\">Link 1</a></div><div class=\\"MuiBox-root MuiBox-root-21\\"><a class=\\"MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorInherit\\" href=\\"/\\">Link 2</a></div><div class=\\"MuiBox-root MuiBox-root-22\\"><a class=\\"MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorInherit\\" href=\\"/\\">Link 3</a></div></div></div></div></div></footer>"`;
